followthemoney/proxy.py

```diff
@@ -264,11 +264,11 @@ def __len__(self):
         return self._size
 
     def __hash__(self):
-        return hash(self.id)
+        return hash(self.id or id(self))
 
     def __eq__(self, other):
         try:
-            return self.id == other.id
+            return self.id == other.id and not (self.id is None and other.id is None)
```

Comment:
The overrides for `__eq__` and `__hash__` make me a bit nervous. The one for `__eq__` is in particular risky. Can we maybe override these in the subclass instead and then make explicit reference to `.parts`?

Reply:
These changes were made so that we can have a set of `EntityProxy`'s. As it stands, the current implementation seems flawed because two objects with completely different data but no ID set will be considered the same. 

ie; without the change:

```python
In [1]: from followthemoney import model
RDFLib Version: 5.0.0

In [2]: a = model.make_entity('Person'); a.add("name", "micha")

In [3]: b = model.make_entity('Person'); b.add("name", "pudo")

In [4]: set([a, b])
Out[4]: {<E(None,'micha')>}
```

with the change:

```python
In [1]: from followthemoney import model
RDFLib Version: 5.0.0

In [2]: a = model.make_entity('Person'); a.add("name", "micha")

In [3]: b = model.make_entity('Person'); b.add("name", "pudo")

In [4]: set([a, b])
Out[4]: {<E(None,'micha')>, <E(None,'pudo')>}
```

I'm fine changing this back if you are concerned, but the MultiPartProxy implementation will need some re-tuning or some documentation around this counter intuitive behavior.